<a href="https://colab.research.google.com/github/JayShekhavat/Diabetes-Prediction-ML-Project/blob/main/Diabetes_ML_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Import all libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

#**Load dataset in csv formate**

In [ ]:
df = pd.read_csv("/content/diabetes_prediction_dataset.csv")

# **Show top 5 records**

In [ ]:
df.head()

gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

#**Show 5 sample records**

In [ ]:
df.sample(5)

gender    age  hypertension  heart_disease smoking_history    bmi  \
26842  Female  27.00             0              0           never  21.33   
26808  Female  25.00             0              0           never  18.01   
85930    Male  75.00             0              0           never  29.95   
24224  Female   0.72             0              0         No Info  19.36   
69300  Female  61.00             0              0           never  27.32   

       HbA1c_level  blood_glucose_level  diabetes  
26842          6.1                  155         0  
26808          5.7                   85         0  
85930          4.0                  159         0  
24224          6.5                  140         0  
69300          5.7                  130         1

In [ ]:
df.columns

Index(['gender', 'age', 'hypertension', 'heart_disease', 'smoking_history',
       'bmi', 'HbA1c_level', 'blood_glucose_level', 'diabetes'],
      dtype='object')

#**Show columns in dataset**

#**Show shape of dataset**

In [ ]:
df.shape

(100000, 9)

#**Show all statistics values of each columns of dataset**

In [ ]:
df.describe()

age  hypertension  heart_disease            bmi  \
count  100000.000000  100000.00000  100000.000000  100000.000000   
mean       41.885856       0.07485       0.039420      27.320767   
std        22.516840       0.26315       0.194593       6.636783   
min         0.080000       0.00000       0.000000      10.010000   
25%        24.000000       0.00000       0.000000      23.630000   
50%        43.000000       0.00000       0.000000      27.320000   
75%        60.000000       0.00000       0.000000      29.580000   
max        80.000000       1.00000       1.000000      95.690000   

         HbA1c_level  blood_glucose_level       diabetes  
count  100000.000000        100000.000000  100000.000000  
mean        5.527507           138.058060       0.085000  
std         1.070672            40.708136       0.278883  
min         3.500000            80.000000       0.000000  
25%         4.800000           100.000000       0.000000  
50%         5.800000           140.000000       0.000000  
75%         6.200000           159.000000       0.000000  
max         9.000000           300.000000       1.000000

#**Show informarion of dataset**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender               100000 non-null  object 
 1   age                  100000 non-null  float64
 2   hypertension         100000 non-null  int64  
 3   heart_disease        100000 non-null  int64  
 4   smoking_history      100000 non-null  object 
 5   bmi                  100000 non-null  float64
 6   HbA1c_level          100000 non-null  float64
 7   blood_glucose_level  100000 non-null  int64  
 8   diabetes             100000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 6.9+ MB


#**Check null values**

In [ ]:
df.isnull().sum()

gender                 0
age                    0
hypertension           0
heart_disease          0
smoking_history        0
bmi                    0
HbA1c_level            0
blood_glucose_level    0
diabetes               0
dtype: int64

###There is no any null values in dataset

#**Check duplicate values in dataset**

In [ ]:
df.duplicated().sum()

3854

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.duplicated().sum()

0

###There is no any duplicate values in dataset

In [ ]:
df.sample(5)

gender   age  hypertension  heart_disease smoking_history    bmi  \
59079  Female  56.0             0              0         No Info  31.94   
96794    Male  21.0             0              0           never  27.32   
97833    Male  80.0             0              1            ever  28.77   
59752    Male  64.0             0              0           never  29.30   
25373  Female  41.0             0              0         current  25.88   

       HbA1c_level  blood_glucose_level  diabetes  
59079          5.0                  200         0  
96794          4.8                   80         0  
97833          6.2                  300         1  
59752          5.0                  160         0  
25373          6.2                  160         0

In [ ]:
df.value_counts().sum()

96146

In [ ]:
df['gender'].value_counts()

gender
Female    56161
Male      39967
Other        18
Name: count, dtype: int64

In [ ]:
df = df[df['gender'] != 'Other']

In [ ]:
df['gender'] = np.where(df['gender'] == 'Male', 1, 2)

<ipython-input-68-adb01bdb2839>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = np.where(df['gender'] == 'Male', 1, 2)


In [ ]:
df = df.drop(['smoking_history'], axis=1)

In [ ]:
df['gender'].value_counts()

gender
2    56161
1    39967
Name: count, dtype: int64

#**Split dataset into train test**

In [ ]:
df.head()

gender   age  hypertension  heart_disease    bmi  HbA1c_level  \
0       2  80.0             0              1  25.19          6.6   
1       2  54.0             0              0  27.32          6.6   
2       1  28.0             0              0  27.32          5.7   
3       2  36.0             0              0  23.45          5.0   
4       1  76.0             1              1  20.14          4.8   

   blood_glucose_level  diabetes  
0                  140         0  
1                   80         0  
2                  158         0  
3                  155         0  
4                  155         0

In [ ]:
df.rename(columns = {'gender':'Gender', 'age': 'Age', 'hypertension':'Hypertension', 'heart_disease':'HeartDisease', 'bmi':'BMI','HbA1c_level':'HBA', 'blood_glucose_level': 'Glucose', 'diabetes':'Diabetes'}, inplace=True)

In [ ]:
df.head()

Gender   Age  Hypertension  HeartDisease    BMI  HBA  Glucose  Diabetes
0       2  80.0             0             1  25.19  6.6      140         0
1       2  54.0             0             0  27.32  6.6       80         0
2       1  28.0             0             0  27.32  5.7      158         0
3       2  36.0             0             0  23.45  5.0      155         0
4       1  76.0             1             1  20.14  4.8      155         0

In [ ]:
x = df.iloc[:,0:-1]
y = df.iloc[:,-1]

In [ ]:
x.head()

Gender   Age  Hypertension  HeartDisease    BMI  HBA  Glucose
0       2  80.0             0             1  25.19  6.6      140
1       2  54.0             0             0  27.32  6.6       80
2       1  28.0             0             0  27.32  5.7      158
3       2  36.0             0             0  23.45  5.0      155
4       1  76.0             1             1  20.14  4.8      155

In [ ]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Diabetes, dtype: int64

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=40)

In [ ]:
y_train.head()

61808    1
34315    0
17195    0
73096    0
41017    0
Name: Diabetes, dtype: int64

In [ ]:
x_test.head()

Gender   Age  Hypertension  HeartDisease    BMI  HBA  Glucose
39009       1  50.0             0             0  31.12  6.6      155
75647       2  17.0             0             0  27.32  6.6      159
27885       2  80.0             0             0  17.76  4.5      159
20837       2  12.0             0             0  24.04  6.1      159
57294       2  41.0             0             0  21.76  6.0      126

In [ ]:
y_test.head()

39009    0
75647    0
27885    0
20837    0
57294    0
Name: Diabetes, dtype: int64

In [ ]:
scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.fit_transform(x_test)

##**K-nearest neighbor Algorithm**

In [ ]:
tree_model = KNeighborsClassifier(n_neighbors=11, p=2, metric='euclidean')
ml_tree_model = tree_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_tree_model.predict(x_test_scaled)

In [ ]:
tree_accuracy = accuracy_score(y_predicted, y_test)


In [ ]:
tree_accuracy

0.9635909705606991

###K-nearest neighbors algorithm give **0.9636** accuracy

##**Logistic regression algorithm**

In [ ]:
lr_model = LogisticRegression()
ml_lr_model = lr_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_lr_model.predict(x_test_scaled)
lr_accuracy = accuracy_score(y_predicted, y_test)
lr_accuracy

0.959221887027983

###Logistic regression algorithm give **0.9593** accuracy

##**Support Vector Machine algorithm**

In [ ]:
svc_model = SVC()
ml_svc_model = svc_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_svc_model.predict(x_test_scaled)
svc_accuracy = accuracy_score(y_predicted, y_test)
svc_accuracy

0.963486944762301

###Support vector machine give **0.9635** accuracy

##**Decision Trees algorithm**

In [ ]:
tree_model = DecisionTreeClassifier()
ml_tree_model = tree_model.fit(x_train_scaled, y_train)

In [ ]:
y_predicted = ml_tree_model.predict(x_test_scaled)
tree_accuracy = accuracy_score(y_predicted, y_test)
tree_accuracy

0.9497035264745657

###Decision tree classifier algorithm give **0.9498** accuracy

In [ ]:
import pickle

In [ ]:
from google.colab import files
with open('ml_tree_model.pkl', 'wb') as file:
  pickle.dump(ml_tree_model, file)

In [ ]:
files.download('/content/ml_tree_model.pkl')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

##**Conclusion**

###I used 4 algorithms, out of these algorithms Logistic regression algorithm give good accuracy than others algorithms.